alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/atoms/RuleAtom.java

```diff
@@ -43,27 +38,92 @@
 import at.ac.tuwien.kr.alpha.commons.terms.Terms;
 import at.ac.tuwien.kr.alpha.core.rules.CompiledRule;
 
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static at.ac.tuwien.kr.alpha.commons.util.Util.oops;
+
 /**
- * Atoms corresponding to rule bodies use this predicate, first term is rule number,
- * second is a term containing variable substitutions.
+ * Atoms corresponding to rule bodies use this predicate, its only term is a Java object linking to the non-ground rule and grounding substitution.
  */
 public class RuleAtom implements Atom {
-	public static final Predicate PREDICATE = Predicates.getPredicate("_R_", 2, true, true);
+	public static final Predicate PREDICATE = Predicates.getPredicate("_R_", 1, true, true);
 
-	private final List<ConstantTerm<String>> terms;
+	private final List<ConstantTerm<RuleAtomData>> terms;
 
-	private RuleAtom(List<ConstantTerm<String>> terms) {
-		if (terms.size() != 2) {
-			throw new IllegalArgumentException();
+	public static class RuleAtomData implements Comparable<RuleAtomData> {
+		private final CompiledRule nonGroundRule;
+		private final Substitution substitution;
+		private final int substitutionHash;
+
+		RuleAtomData(CompiledRule nonGroundRule, Substitution substitution) {
+			this.nonGroundRule = nonGroundRule;
+			this.substitution = substitution;
+			this.substitutionHash = substitution.hashCode();
 		}
 
-		this.terms = terms;
+		@Override
+		public int compareTo(RuleAtomData other) {
+			if (nonGroundRule.getRuleId() != other.nonGroundRule.getRuleId()) {
+				return Integer.compare(nonGroundRule.getRuleId(), other.nonGroundRule.getRuleId());
+			}
+			if (substitution.getSubstitution().size() != other.getSubstitution().getSubstitution().size()) {
+				throw oops("RuleAtoms over the same rule have different-sized substitutions.");
+			}
+			// Note: We assume here that substitutions for the same nonGroundRule are all over the same variables and iteration order.
```

Comment:
In addition to having to rely on this assumption, there is also never a point where a connection is made from the variables set in the `Substitution`s vs. the ones actually occurring in the rule. See PR #354 for a suggestion of a more robust version that also doesn't make this assumption.

Reply:
Thank you for pointing at this issue and suggesting a solution to it. After some deliberation, I changed the code in another way for the following reasons:
- The suggested code repeatedly collects all occurring variables of the underlying non-ground rule (every time `compareTo` is called).
- If we really want to ensure that a given `Substitution` fits the `CompiledRule`, it should be done in the constructor and not on a call to `compareTo`.
- The `RuleAtom` is in the core package, so the condition that substitutions fit the non-ground rule should never be violated.
- The assumption mentioned in the comment above, about variable iteration order, is actually always true as the `Substitution` interface guarantees that the substitution is backed by a TreeMap, which has a well-defined iteration order.
- The other part of the assumption (that substitutions are over the same variables) can easily be checked. The changed code does that now.

Please have a look at the updated code and feel free to criticise my reasoning above.